flang: update support and add flang-rt

Acture opened the pull request:
Supersedes #391942, and should close #516651.

## Summary
Package LLVM Flang as a first-class Fortran compiler in nixpkgs, with runtime library, cc-wrapper integration, and driver compatibility patches backported from upstream.

Based on the excellent initial implementation by @picostove.

## Key Changes

### Packages:

- Update the flang package with standalone build, passthru metadata (langFortran, isFlang, isClang, hardeningUnsupportedFlags)
- Add flang-rt runtime library (libFortranRuntime, libFortranDecimal)

### cc-wrapper integration:

- Recognize -fc1 as frontend invocation (analogous to clang's -cc1)
- Strip -frandom-seed=* when wrapping flang (injected by reproducible-build hook; flang doesn't accept it)
- Use existing hardeningUnsupportedFlags to disable unsupported hardening flags

### Driver patches (backported from upstream):

- warn-on-fbuiltin-and-fno-builtin — accept -fbuiltin/-fno-builtin with warning instead of hard error
- accept-and-ignore-some-gfortran-optimization-flags — handle -fexpensive-optimizations etc.
- use-xflang-in-diagnostics — suggest -Xflang instead of -Xclang in error messages

### MLIR: fix MLIRConfig.cmake for standalone builds

MLIRConfig.cmake unconditionally overwrites `MLIR_TABLEGEN_EXE`, breaking standalone
builds that provide their own pre-built `mlir-tblgen` (e.g. in Nix sandboxed builds).
The patch adds guards to respect caller-set values and auto-creates an imported
`mlir-tblgen` target for downstream consumers. This replaces the previous dummy target
workaround in flang's CMakeLists.txt.

Upstream issue: https://github.com/llvm/llvm-project/issues/150986

### flang-rt: remove unused dependencies

flang-rt does not link against MLIR or libclang; removed unused `CLANG_DIR`/`MLIR_DIR`
cmake flags, build inputs, and the empty `dev` output.

## Tests:

- compile-smoke — basic compilation and response file (@file) support
- driver-flags — wrapper flag filtering, driver diagnostics, and regression coverage for #452306 (comment) issues

## Design
No user-supplied flags are silently dropped. Flag incompatibility is handled at three layers:

1. **cc-wrapper** — flags that should never reach flang are not injected in the first place (hardeningUnsupportedFlags, -frandom-seed filtering)
2. **Upstream patches** — flags commonly passed by Fortran build systems produce warnings, not hard errors
3. **Passthru metadata** — downstream tooling can detect flang and adapt accordingly

## Test plan
- [x] `nix-build -A llvmPackages_21.flang`
- [x] `nix-build -A llvmPackages_21.flang-rt`
- [x] `nix-build -A llvmPackages_21.flang.passthru.tests.compile-smoke`
- [x] `nix-build -A llvmPackages_21.flang.passthru.tests.driver-flags`
- [ ] @VincentVanlaer's reproducer (basic Fortran compilation via mkDerivation)

## Checklist
- [x] I have read the contributing guidelines.
- [x] The PR is formatted using `nix fmt`.

Acture commented:
Build took an extremely long time. wasn't expecting this though

RossComputerGuy commented:
I think I just ran into a bug with nixpkgs-review, the build is still running.

alyssais commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -707,7 +712,13 @@ stdenvNoCC.mkDerivation {
         }" "${libc_dev}${libc.incdir or "/include"}" >> $out/nix-support/libc-cflags
       ''
-      + optionalString isGNU ''
+      + optionalString isGNU ''''
```
What's up with this empty string?

Acture replied:
Fixed, it was a mistake.

alyssais commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -400,8 +400,21 @@ let
+      flang-unwrapped = callPackage ./flang {
         mlir = tools.mlir;
+        # The C++ template instantiations in flang frequently cause g++ to be
+        # OOM-killed, so use clang++ instead.
```
hmm, on a machine with how much memory?

Acture replied:
This comment was originally from @picostove, I'm not sure about the exact memory specs. I've encountered similar g++ OOM issues with olberg bot in a past flang-related PR though.
But I think it's better to remove this override as of now.

alyssais reviewed (commented):
I think this should be split into multiple commits.  There's a lot going on here.

alyssais commented:
Perhaps adding flang-rt could also be a commit of its own?

Acture commented:
I believe the previous build issue was fixed by the latest version(d2ad9b4). Followed the way of master instead of staging branch.
Is there any way to extend the build time for ci so it doesn’t get cancelled? @alyssais @RossComputerGuy 
Or what else should I do to help move this forward?

Acture commented:
> > > > Is there any way to extend the build time for ci so it doesn’t get cancelled?
> > > 
> > > 
> > > For ofborg, there's a way but I haven't seen much success in people asking for the time to be raised.
> > 
> > 
> > Do you think it's better to leave cc-wrapper untouched for now? So envs and most pkgs don't have to be rebuilt
> 
> No, it's fine if things timeout because Ofborg isn't reliable. We mostly care if it builds on Hydra and doesn't cause a channel blocker.

Is there anything that I could do to help move this forward?

emilazy reviewed (dismissed):
Just a matter of limited review capacity on the LLVM team. I’m interested in this work, though, as I’d like to use it by default on Darwin instead of GFortran if possible.

Rebuilds are fine – we rebuild the world every few weeks – but this should target `staging` as a result. You’ll need to rebase, as there are changes on `staging` relating to the C++ include stuff. (I’d suggest waiting a couple of days for https://github.com/NixOS/nixpkgs/pull/445668, which will cause conflicts in `pkgs/development/compilers/llvm/common/default.nix` anyway.)

Acture commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -568,10 +568,15 @@ stdenvNoCC.mkDerivation {
     ln -sv ${targetPrefix}gfortran $out/bin/${targetPrefix}f77
     export named_fc=${targetPrefix}gfortran
+      elif [ -e $ccPath/flang ]; then
+        wrap ${targetPrefix}flang $wrapper $ccPath/flang
```
I think fc coud be a good name.

Acture commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -751,14 +756,17 @@ stdenvNoCC.mkDerivation {
-    ''
+    +
+      optionalString
+        (libcxx == null && isClang && (cc.langCC or false) && (useGccForLibs && gccForLibs.langCC or false))
```
I think `gccForLibs` probably comes from previous code from `staging`, so I left it there.

Acture commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -819,6 +827,7 @@ stdenvNoCC.mkDerivation {
           && !(cc.isROCm or false)
           && !targetPlatform.isDarwin
           && !targetPlatform.isAndroid
+          && ((cc.langC or false) || (cc.langCC or false))
```
This conditional logic for  -nostdlibinc was inherited from @picostove's earlier work. I wasn't entirely certain of the original rationale for excluding GCC (or Darwin/ROCm), so I left if as-is. `monorepoSrc != null` and `libcxx ` likewise

Will remove.

picostove commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -396,9 +400,19 @@ let
         mlir = tools.mlir;
       };
+      flang = wrapCCWith rec {
+        inherit (buildLlvmTools.clang) libcxx;
```
I don't recall a good reason for this. I think this is also the wrong package set anyway?

Maybe I intended for it to allow `flang` to link executables against `libc++` without additional config.

Acture commented:
I suppose current CI failure caused by PIE warning is unrelated to this work?
```
       > evaluation warning: The 'pie' hardening flag has been removed in favor of enabling PIE by default in compilers and should no longer be used. PIE can be disabled with the -no-pie compiler flag, but this is usually not necessary as most build systems pass this if needed. Usage of the 'pie' hardening flag will become an error in future.
       > evaluation warning: The 'pie' hardening flag has been removed in favor of enabling PIE by default in compilers and should no longer be used. PIE can be disabled with the -no-pie compiler flag, but this is usually not necessary as most build systems pass this if needed. Usage of the 'pie' hardening flag will become an error in future.
       > Chunk 00 on aarch64-darwin done [17217616KB max resident, 0m 35.94ss elapsed] nix-env -f /nix/store/1d8x5bzd3w4kfnlgsdrx8cnyp2j1np0h-source/ci/eval/chunk.nix --eval-system aarch64-darwin --option restrict-eval true --option allow-import-from-derivation false --query --available --out-path --json --show-trace --arg chunkSize 8000 --arg myChunk 00 --arg attrpathFile /nix/store/5q7574hah453aw5wiin8530hg75zfq0l-attrpaths-superset.json/paths.json --arg systems [ "aarch64-darwin" ] --arg includeBroken false -I /nix/store/1d8x5bzd3w4kfnlgsdrx8cnyp2j1np0h-source -I /nix/store/5q7574hah453aw5wiin8530hg75zfq0l-attrpaths-superset.json/paths.json
       > Nixpkgs on aarch64-darwin evaluated with warnings, aborting
       For full logs, run:
         nix log /nix/store/8psblsbp85q25mib9pb9rvb0ywgcpjhb-nixpkgs-eval-aarch64-darwin.drv
Error: Process completed with exit code 1.
```

alyssais reviewed (commented):
Otherwise looking good.

philiptaron commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -568,10 +568,15 @@ stdenvNoCC.mkDerivation {
   ''
 
   + optionalString cc.langFortran or false ''
+      if [ -e $ccPath/${targetPrefix}gfortran ]; then
```
indent is odd here

Acture replied:
Fixed.

philiptaron reviewed (approved):
Non-blocking style comment. Is there a representative Fortran package I could build with `flang`, beyond just building the compiler?

Acture commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -44,7 +44,7 @@ while (( "$n" < "$nParams" )); do
     case "$p" in
         -[cSEM] | -MM) dontLink=1 ;;
-        -cc1) cc1=1 ;;
+        -cc1 | -fc1 ) cc1=1 ;;
```
> Research: the `-fc1` flag in a flang command line tells flang to run in "frontend-only" mode, which is the compiler's internal frontend interface.
> 
> Docs: https://flang.llvm.org/docs/FlangDriver.html
> 
> It's the exact analogue of `-cc1` in `clang`.

Do you think I should add a comment here to explain?

philiptaron replied:
Nah, I just didn't know. Now that `git blame` will point to this commit, it'll also point to this review.

Acture commented:
let me know if there's anything else I could do to move this forward.

VincentVanlaer commented:
Building the following derivation gives me the error. Swapping flang by gfortran produces the correct binary.

<details>
<summary>Reproducer</summary>

```nix
let
  unstable = fetchTarball {
    url = "https://github.com/nixos/nixpkgs/archive/c6245e83d836d0433170a16eb185cefe0572f8b8.tar.gz";
    sha256 = "1dvhyaddi7d4fkj63hns1xrdqcmyyq24y89k0cdwxs8s3619bx8v";
  };
  pkgs = import unstable { };
  pkgs-patched = import
    (pkgs.applyPatches
      {
        src = unstable;
        name = "update-flang";
        patches = [
          (
            builtins.fetchurl {
              url = "https://github.com/NixOS/nixpkgs/compare/177d9ea796371764c5938ac9d2c1ee884495cae2...e2d1358f1beaa0bae454b21036f9bdc940919488.patch";
              sha256 = "0fgmsh9yqwhiarqjs88prc1r7cf8rdpclbzrdr03sm4nqbyv989v";
            })
        ];
      })
    { };
in
pkgs.stdenvNoCC.mkDerivation {
  name = "hello";
  phases = [ "buildPhase" "installPhase" ];
  nativeBuildInputs = [ pkgs-patched.llvmPackages.flang ];

  buildPhase = /* bash */ ''
    cat >hello.f90 <<EOF
    program main
      write(*,*) "Hello world"
    end program main
    EOF
    flang -o hello hello.f90
  '';

  installPhase = ''
    mkdir -p $out/bin
    cp hello $out/bin
  '';
}
```
</details>

Acture commented:
> Building the following derivation gives me the error. Swapping flang by gfortran produces the correct binary.
> 
> Reproducer

I have fixed the flag incompatibility using a dedicated wrapper script. Please feel free to test your reproducer again!

Currently, natively disabling -frandom-seed and other C-centric flags in cc-wrapper is quite messy and would result in too many hard-coded conditionals globally. Given Flang's status, it is likely, IMHO ,that the driver will gracefully handle these standard hardening flags in the near future.

Perhaps we should consider a separate PR to structurally refactor cc-wrapper later, but for the time being, I think localized wrapper script seems to be the best choice

alyssais commented on pkgs/development/compilers/llvm/common/default.nix:
Won't this do the wrong thing when cross compiling?  It'll be the build platform's shell.


```suggestion
            #!${runtimeShell}
```

alyssais commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -479,14 +479,96 @@ makeScopeWithSplicing' {
+        buildCommand = ''
+            mkdir -p $out/bin
+            cat > $out/bin/flang-new <<'EOF'
+            #!${stdenv.shell}
```
Would be a lot easier to read if this were a separate .sh file rather than all inline like this.

Acture committed:
clang, cc-wrapper: make wrapper flang-aware

Build on the initial LLVM Flang wrapper work by exposing the needed clang passthru and teaching cc-wrapper about flang-specific flag handling.

Co-authored-by: stove <stove@rivosinc.com>
Co-authored-by: acture <acture@gmail.com>
Co-authored-by: Emily <hello@emily.moe>
Co-authored-by: Philip Taron <philip.taron@gmail.com>

Acture committed:
flang-rt: add runtime package

Add a standalone derivation for LLVM's flang-rt runtime libraries
(libFortranRuntime, libFortranDecimal). flang-rt is a runtimes-style
package: building it requires running a working flang binary at *build*
time, but the produced libraries are linked into derivations targeting
the *host*.

The runtime is therefore wired up using `buildFlang`, sourced from
`buildLlvmPackages.flang-unwrapped` so that the build-platform flang
(spliced via the LLVM package set) compiles the runtime, while the
resulting libraries are still produced for the host platform. This
mirrors the existing `buildLlvmTools.tblgen` / `buildLlvmPackages.tblgen`
pattern used elsewhere in the LLVM package set.

`unwrapped` rather than the wrapped flang is used because flang-rt's
CMake glue invokes the compiler directly and does not need (or want)
Nix's cc-wrapper resource-dir and rpath plumbing during the runtime
build.

flang-rt only consumes LLVM's CMake helpers and the build-time flang;
it does not link against MLIR or libclang, so neither MLIR_DIR nor
CLANG_DIR is passed and the unused empty `dev` output is omitted.

Co-authored-by: stove <stove@rivosinc.com>
Co-authored-by: acture <acture@gmail.com>

Acture committed:
flang: package standalone flang with LLVM 20+ driver fixes

Wire flang into llvmPackages as a first-class Fortran compiler. flang
is built standalone on top of the LLVM/clang package set rather than
bundled into the llvm derivation, with passthru metadata
(`langFortran`, `isFlang`, `isClang`, `hardeningUnsupportedFlags`) so
cc-wrapper and downstream tooling can detect and adapt to it.

Driver compatibility patches backported from upstream are applied
selectively per LLVM version:

  * `use-xflang-in-diagnostics` is applied to LLVM 20 and newer; it
    teaches the driver to suggest `-Xflang` instead of `-Xclang` in
    error messages for options only available to `flang -fc1`.
  * `warn-on-fbuiltin-and-fno-builtin` and
    `accept-and-ignore-some-gfortran-optimization-flags` are applied
    to LLVM 20 and 21 only. LLVM 22 has equivalent behaviour merged
    upstream (`warn_drv_invalid_argument_for_flang` and
    `clang_ignored_gcc_optimization_f_Group` handling in
    clang/lib/Driver/ToolChains/Flang.cpp), so the patches are skipped
    there.

Patches live under pkgs/development/compilers/llvm/21/flang/ and are
shared across versions via patches.nix. They are applied to a private
libclang variant rather than the flang source tree because standalone
flang resolves driver/option definitions through the installed libclang
package.

Two focused passthru tests are added:

  * `compile-smoke` exercises basic compilation and `@response-file`
    handling.
  * `driver-flags` covers wrapper flag isolation
    (`NIX_CFLAGS_COMPILE` must not leak into flang;
    `NIX_FFLAGS_COMPILE` must reach it), the backported driver
    diagnostics, and regression coverage for previously hard-erroring
    flags.

Build on the earlier standalone flang work by @picostove.

Co-authored-by: stove <stove@rivosinc.com>
Co-authored-by: acture <acture@gmail.com>
Co-authored-by: Alyssa Ross <hi@alyssa.is>

Acture committed:
llvm/mlir: fix MLIRConfig.cmake to support external tablegen overrides

MLIRConfig.cmake unconditionally overwrites MLIR_TABLEGEN_EXE and does
not create an imported target, breaking standalone builds that provide
their own mlir-tblgen binary (e.g. Nix sandboxed builds).

The patch adds guards to respect caller-set MLIR_TABLEGEN_EXE,
MLIR_PDLL_TABLEGEN_EXE, and MLIR_SRC_SHARDER_TABLEGEN_EXE values, and
auto-creates an imported mlir-tblgen target for downstream consumers.
This replaces the previous dummy-target workaround in flang's
CMakeLists.txt.

Upstream issue: https://github.com/llvm/llvm-project/issues/150986

Co-authored-by: Alexander Bantyev <balsoft@balsoft.ru>

alyssais commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -218,6 +218,27 @@ if [[ -e @out@/nix-support/add-local-cc-cflags-before.sh ]]; then
+if [ "@isFlang@" = 1 ]; then
+    # Nix's reproducible-build hook injects -frandom-seed=..., but flang does
+    # not currently accept that option. Strip it here so the wrapper remains
+    # deterministic without surfacing a spurious flang driver failure.
```
This is still silently stripping an argument.  Is it right for the wrapper to be looking at NIX_CFLAGS_COMPILE at all?  Shouldn't it use NIX_FFLAGS_COMPILE or something?  We already have NIX_GNATFLAGS_COMPILE as well, for precedent.

Acture replied:
Switched.

alyssais commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -497,7 +497,7 @@ makeScopeWithSplicing' {
         };
         flangRt = callPackage ./flang-rt {
-          flang-unwrapped = flangUnwrapped;
+          buildFlang = buildLlvmPackages.flang-unwrapped;
```
Would be good for this commit to have a bit more explanation.

Acture replied:
Previous flang-unwrapped = flangUnwrapped passed the host-stage flang. flang-rt's CMake invokes flang during the configure/build phase to compile the runtime sources, and the host-stage binary is unrunnable on the build machine in any cross scenario (e.g. pkgsCross.aarch64-multiplatform.llvmPackages_21.flang-rt on an x86_64 builder), so the configure step fails before the first source file is compiled. On native builds it happens to work because host == build.

buildLlvmPackages.flang-unwrapped resolves to the build-platform projection of the LLVM scope — same splicing pattern already used for buildLlvmTools.tblgen / buildLlvmPackages.tblgen. CMake gets a binary that runs on the builder, while the produced libraries still target the host triple via LLVM_DEFAULT_TARGET_TRIPLE and the cross flags carried in by stdenv.

alyssais commented on pkgs/development/compilers/llvm/common/mlir/default.nix:
```diff
@@ -39,6 +39,13 @@ stdenv.mkDerivation (finalAttrs: {
+    # mlir-tblgen target for downstream consumers. This replaces the previous dummy target
+    # workaround in flang's CMakeLists.txt.
+
+    # Upstream issue: https://github.com/llvm/llvm-project/issues/150986
```
Please squash this into the commit that added it the patch.

Acture replied:
Squashed.

alyssais reviewed (commented):
This looks so much better.  Thanks a lot for reworking it.  I don't think it's far off now.

Acture commented:
Also reorganized into 4 atomic commits:
- `clang, cc-wrapper: make wrapper flang-aware`
- `flang-rt: add runtime package` (wired with `buildFlang` from the start)
- `flang: package standalone flang with LLVM 20+ driver fixes` — patches now apply to LLVM 20+; the two LLVM 22 already merged upstream are skipped there
- `llvm/mlir: fix MLIRConfig.cmake to support external tablegen overrides` — also rolls in the dummy-target removal it obsoletes

@alyssais  Take another look when you have time.

alyssais reviewed (approved):
I did not expect it could turn out so nicely.  Thank you!

alyssais commented:
I'll leave it a bit to give other people a chance for last comments.  Ping me in a week if it looks ready for merge then. :)
